common-npm-packages/codeanalysis-common/Common/SpotbugsTool.ts

```diff
@@ -0,0 +1,144 @@
+            }
+            toolRunner.arg(['check']);
+        }
+        return toolRunner;
+    }
+
+    protected getSpotBugsGradlePluginVersion(): string {
+        let userSpecifiedVersion = tl.getInput('spotbugsGradlePluginVersion');
+        if (userSpecifiedVersion) {
+            const pluginVersion = userSpecifiedVersion.trim();
+            return pluginVersion;
+        } 
+        return defaultPluginVersion;
+    }
+
+    /**
+     * Implementers must specify where the XML reports are located
+     */
+    protected getBuildReportDir(output: ModuleOutput): string {
+        return path.join(output.moduleRoot, 'reports', 'spotbugs');
+    }
+
+    /**
+     * Report parser that extracts the number of affected files and the number of violations from a report
+     *
+     * @returns a tuple of [affected_file_count, violation_count]
+     */
+    protected parseXmlReport(xmlReport: string, moduleName: string): [number, number] {
+        let jsonCounts: [number, number] = [0, 0];
+
+        let reportContent: string = fs.readFileSync(xmlReport, 'utf-8');
+        xml2js.parseString(reportContent, (err, data) => {
+            jsonCounts = SpotbugsTool.parseJson(data);
+            if (jsonCounts == null) {
+                tl.debug(`[CA] Empty or unrecognized SpotBugs XML report ${xmlReport}`);
+                return null;
+            }
+
+            let violationCount: number = jsonCounts[1];
+
+            // No files with violations, return now that it has been marked for upload
+            if (violationCount === 0) {
+                tl.debug(`[CA] A SpotBugs report was found for module '${moduleName}' but it contains no violations`);
+                return null;
+            }
+
+            tl.debug(`[CA] A SpotBugs report was found for for module '${moduleName}' containing ${violationCount} issues - ${xmlReport}`);
+        });
+
+        return jsonCounts;
+    }
+
+    /**
+     * Extracts the number of affected files and the number of violations from a report JSON object
+     * @param data JSON object to parse
+     * @returns a tuple of [affected_file_count, violation_count]
+     */
+    private static parseJson(data: any): [number, number] {
+        // If the file is not XML, or is not from SpotBugs, return immediately
+        if (!data || !data.BugCollection ||
+            !data.BugCollection.FindBugsSummary || !data.BugCollection.FindBugsSummary[0] ||
+            !data.BugCollection.FindBugsSummary[0].PackageStats ||
+            !data.BugCollection.FindBugsSummary[0].PackageStats[0].ClassStats) {
+            return null;
+        }
+
+        let fileCount: number = 0;
+        let violationCount: number = 0;
+
+        // Extract violation and file count data from the sourceFile attribute of ClassStats
+        let filesToViolations: Map<string, number> = new Map(); // Maps files -> number of violations
+        data.BugCollection.FindBugsSummary[0].PackageStats[0].ClassStats.forEach((classStats: any) => {
```

Comment:
I think there is a code duplication. You can create variable and assign it that expression joined via optional operator. You can use it in `if` above.

Reply:
@matskan Good point. Did it.